### PR TITLE
comm port autodetection rewrite

### PR DESCRIPTION
**Pull request by bzed**
Hi :)

So this is my take on fixing #3394.
I *know* that comm port autodetection worked before, but with the current staging/devel checkout it failed, with Python 2 and Python 3.

So I've started to look into the avr code and tried to debug it, got annoyed, and removed the whole code and replaced the whole thing by a stk500v2 sync and sign_on. I don't think there is a need to check if there is actually a programmable board behind that port, also supporting all possible boards sounds impossible.

What I'm absolutely not sure about is the timeout setting, I think it should be done to 1s or something like that.
I've tested whats happening with my multimeter (which does not even accept writes, especially not with !=9k6 baud), Octoprint ignores it just fine.

There is (imho) one bug left: after a disconnect, it is not able to detect the port on the first try. As far as I can see the fix for that is to 'sync' a few times until its in sync. Tiny loop around [0x30].

Otherwise the coew works just fine here.

### CODE REVIEW

File: src/octoprint/util/comm.py [hunk 1]
```diff
@@ -2616,23 +2603,29 @@ def _detect_port(self):
...
+				try:
+				        serial_obj = serial.Serial(str(p), baudrate, timeout=1, writeTimeout=10000)
+				except Exception:
+					self._logger.info("Could not connect to {} ({}).".format(p, baudrate))
+					continue
 				try:
-					self._log("Trying {}".format(p))
-					programmer.connect(p)
-					serial_obj = programmer.leaveISP()
+					if _send_message(serial_obj, [0x30]) != b'\x00\n':
```
Comment (bzed) [inline src/octoprint/util/comm.py]: We should try this up to 5 times. This should fix a reconnect after a  disconnection.

**Comment by bzed** [comment]
Oh well, yes dear bot.
Applies fine to devel, though :)

**Comment by foosel** [comment]
Ignore the bot here, I forgot to whitelist the branch.

I'll take a closer look at this ASAP, but it might be only after my upcoming vacation.

**Comment by foosel** [comment]
Ah, but don't ignore the test failure, though I can also take care of that. Wrapping things up for 2019 now though, so it'll have to wait until I'm back to work in January 🎄

**Comment by bzed** [comment]
> Ah, but don't ignore the test failure, though I can also take care of that. Wrapping things up for 2019 now though, so it'll have to wait until I'm back to work in January

yeah, but I have similar issues with free time, so I thought I'd trow the branch at you in case you want to fix it before I find the time to do so ;)

**Comment by bzed** [comment]
@foosel oh well, that was easy. Travis is happy.

**Comment by bzed** [comment]
Reconnecting after disconnect should work now, too. Not tested yet, though.

File: src/octoprint/util/comm.py (hunk 1 above)
Comment (bzed) [reply src/octoprint/util/comm.py]: Fixed in 08dd797

**Comment by foosel** [comment]
Planning to get rid of the "try to enter programming mode" approach in 1.4.1 altogether, since this is all just horribly messy anyhow. Hence not going to merge this (especially not since I couldn't get it to work on my end with printers available to me as discussed on IRC, so there's something wonky about all of this anyhow).

**Pull request closed without merging**